netcheck/netcheck.go

```diff
@@ -453,12 +454,62 @@ func (c *Client) GetReport(ctx context.Context) (*Report, error) {
 			ret.HairPinning.Set(false)
 		}
 	}
+	mu.Unlock()
 
-	// TODO: if UDP is blocked, try to measure TCP connect times
-	// to DERP nodes instead? So UDP-blocked users still get a
-	// decent DERP node, rather than being randomly assigned to
-	// the other side of the planet? Or try ICMP? (likely also
-	// blocked?)
+	// Try HTTPS latency check if UDP is blocked and all checkings failed
+	if !anyV4() {
+		c.logf("netcheck: UDP is blocked, try HTTPS")
+		var wg sync.WaitGroup
+		for _, server := range stuns4 {
+			if _, ok := ret.DERPLatency[server]; ok {
+				continue
+			}
+			server := server
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				host, _, err := net.SplitHostPort(server)
+				if err != nil {
+					return
+				}
+
+				var result httpstat.Result
+				hctx, cancel := context.WithTimeout(httpstat.WithHTTPStat(context.Background(), &result), 5*time.Second)
+				defer cancel()
+
+				// This is a 404 page, but it should be fine for checking latency
+				u := fmt.Sprintf("https://%s/derp/latency-check", host)
+				req, err := http.NewRequestWithContext(hctx, "GET", u, nil)
+				if err != nil {
+					c.logf(err.Error())
+					return
+				}
+
+				resp, err := http.DefaultClient.Do(req)
+				if err != nil {
+					c.logf(err.Error())
+					return
+				}
+				_, err = io.Copy(ioutil.Discard, resp.Body)
+				if err != nil {
+					c.logf(err.Error())
+					return
+				}
+				resp.Body.Close()
```

Comment:
make this `defer resp.Body.Close()` and put it before the `io.Copy` line so it gets called even if there's an `io.Copy` error.